### PR TITLE
[Java] Fixing the creation of OnnxTensors from scalars, adding tests

### LINKED ISSUE #7358: [JAVA] Cannot create an ONNXTensor of boolean scalar due to exception thrown on reshape
I am trying to pass a boolean scalar (for the dropout op, training mode input), into ORT Java.
Because of the lack of BooleanBuffer, in order to do so I first need to use reshape from OrtUtil like so:
```OrtUtil.reshape(arr, shape)```

before passing it into ONNXTensor.createTensor

This works fine in the general case with more dimensions, but in case of a scalar, if I try using shape [], I get:
```
[info]   java.lang.IllegalArgumentException: Arrays with less than 1 and greater than 8 dimensions are not supported.
[info]   at ai.onnxruntime.OrtUtil.transformShape(OrtUtil.java:28)
[info]   at ai.onnxruntime.OrtUtil.newBooleanArray(OrtUtil.java:86)

```
with shape [1] I get:

```
[info]   java.lang.ClassCastException: [Z cannot be cast to [Ljava.lang.Object;
[info]   at ai.onnxruntime.OrtUtil.reshape(OrtUtil.java:284)
[info]   at ai.onnxruntime.OrtUtil.reshape(OrtUtil.java:179)
```

and with shape [1,1] I get:
```
[info]  ai.onnxruntime.OrtException: Error code - ORT_FAIL - message: Node (-531443106) Op (Dropout) [ShapeInferenceError] training_mode of Dropout must be a scalar.
```

### PR DESCRIPTION
**Pull request by Craigacp**
**Description**:
Previously while you could get a scalar as an output from an ONNX model, you couldn't build one to supply as an input. This PR fixes that, and adds a test for this behaviour (as it did work in the past but was broken at some point).

**Motivation and Context**
- Why is this change required? What problem does it solve? Allows the creation of scalar tensors.
- If it fixes an open issue, please link to the issue here. Fixes #7358.

### CODE REVIEW

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java [hunk 1]
```diff
@@ -314,8 +315,9 @@ public static OnnxTensor createTensor(OrtEnvironment env, Object data) throws Or
   }
 
   /**
-   * Create a Tensor from a Java primitive or String multidimensional array. The shape is inferred
-   * from the array using reflection.
+   * Create a Tensor from a Java primitive, String, primitive multidimensional array or String
+   * multidimensional array. The shape is inferred from the object using reflection. The default
+   * allocator is used.
```
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxTensor.java:320]: Has an argument allocator

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java [hunk 2]
```diff
@@ -345,7 +347,14 @@ static OnnxTensor createTensor(OrtEnvironment env, OrtAllocator allocator, Objec
         }
       } else {
         if (info.shape.length == 0) {
-          data = OrtUtil.convertBoxedPrimitiveToArray(data);
+          data = OrtUtil.convertBoxedPrimitiveToArray(info.type, data);
+          if (data == null) {
+            throw new OrtException(
+                "Failed to convert a boxed primitive to an array, this is an error with ORT itself, please report it. JavaType = "
```
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxTensor.java:353]: >this is an error with ORT itself [](http://example.com/codeflow?start=66&length=32)

Is there a specific message returned from ORT and can it be propagated?

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java [hunk 3]
```diff
@@ -314,8 +315,9 @@ public static OnnxTensor createTensor(OrtEnvironment env, Object data) throws Or
   }
 
   /**
-   * Create a Tensor from a Java primitive or String multidimensional array. The shape is inferred
-   * from the array using reflection.
+   * Create a Tensor from a Java primitive, String, primitive multidimensional array or String
+   * multidimensional array. The shape is inferred from the object using reflection. The default
```
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OnnxTensor.java:319]: >The shape is inferred from the object using reflection. [](http://example.com/codeflow?start=29&length=55)

I think there might be a problem with the statement. This function is about scalars. Scalar shape is usually empty, according the definition, although the length of data calculated is still 1. Is this in agreement with what is being inferred here.

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java (hunk 2 above)
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxTensor.java:353]: Ah, no I mean if you get that exception it's a logic error in the Java API, and they should report it because we added support for a new type and missed this method. It's unlikely to be triggered until Java gets value types and then there will be user created primitives.

File: java/src/main/java/ai/onnxruntime/OrtUtil.java [hunk 4]
```diff
@@ -383,15 +383,48 @@ private static void flattenString(Object[] input, ArrayList<String> output) {
...
-   * @return The boxed primitive in an array.
+   * @return The primitive in an array.
    */
-  static Object convertBoxedPrimitiveToArray(Object data) {
-    Object array = Array.newInstance(data.getClass(), 1);
-    Array.set(array, 0, data);
-    return array;
+  static Object convertBoxedPrimitiveToArray(OnnxJavaType javaType, Object data) {
+    switch (javaType) {
+      case FLOAT:
```
Comment (yuslepukhin) [inline java/src/main/java/ai/onnxruntime/OrtUtil.java:396]: >case FLOAT: [](http://example.com/codeflow?start=6&length=11)

Still no support for FLOAT16 and BFLOAT16? We introduced it in C#.

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java (hunk 1 above)
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxTensor.java:320]: Ooops, I copied that over from the other one. I'll fix it.

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java (hunk 2 above)
Comment (yuslepukhin) [reply java/src/main/java/ai/onnxruntime/OnnxTensor.java:353]: What I meant, it is hard to report something that has no starting point for investigation on our side, or, perhjaps, given a message, the user may figure it out on their own.

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java (hunk 3 above)
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxTensor.java:319]: Yes it is. It infers the shape of scalars to be an empty array, which seems to be how it's represented in the native library (though I only figured that out by tracing the CreateTensorAsOrtValue function in the C API).

File: java/src/main/java/ai/onnxruntime/OrtUtil.java (hunk 4 above)
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OrtUtil.java:396]: Not as part of this PR. C# has value types so there is a sensible thing to represent it. In Java we could use a `short` and it'll be unpleasant to use, or we can use some reference type and it'll be slower.

I'm considering adding a way to pass in a `float` or `FloatBuffer` and have the native code downcast that to fp16 or bfloat16 as an input. That would be an alternative that wouldn't necessarily be any slower, but it would be ugly if users are taking the output of one model that's fp16 and passing it into another.

Basically there's no really good way of doing this in Java yet, all the options are bad in one way or another. We can pick a way to support and then note that it might evolve as the Java platform changes, or just not support it until the platform has better functionality.

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java (hunk 2 above)
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxTensor.java:353]: If they paste in the stack trace and associated message then I'll know what went wrong. That's why it prints the javatype and data value at the end of the exception message.

File: java/src/main/java/ai/onnxruntime/OrtUtil.java (hunk 4 above)
Comment (yuslepukhin) [reply java/src/main/java/ai/onnxruntime/OrtUtil.java:396]: Here is my suggestion. Pass it as primitive short type array via separate entry point(s) that would telll ORT C API that this, in fact, is a float16/bfloat16.

Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OrtUtil.java:396]: Sure. I'd prefer to do that in a separate PR though. This one is quite narrowly focused on fixing the bug, whereas adding support for fp16 is going to hit a lot of different places.

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java (hunk 1 above)
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxTensor.java]: Fixed.

File: java/src/main/java/ai/onnxruntime/OnnxTensor.java (hunk 2 above)
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxTensor.java]: I updated the message slightly to say it's a bug in the Java API and they should report the exception message.

**Comment by Craigacp** [comment]
Could someone trigger the remaining pipelines? I think this is good to go once those are done?

**Comment by Craigacp** [comment]
Looks like everything has passed. Is there anything else to do with this PR before merging?

**Review by yuslepukhin — APPROVED** [review]
:shipit:
